accepted/0000-caasp-integration.md

```diff
@@ -0,0 +1,106 @@
+- Document how to deploy CaaS Platform using AutoYaST
+
+# Detailed design
+[design]: #detailed-design
+
+## Node status reporting
+
+All the nodes in the cluster can be registered to Uyuni/SUSE Manager using [bootstrapping methods](https://opensource.suse.com/doc-susemanager/suse-manager/client-configuration/registration-overview.html).
+This brings two major benefits:
+
+- The user can see from Uyuni/SUSE Manager if a node has packages updates and patch available, along with all the information that a Salt minion is presenting when registered to Uyuni/SUSE Manager (e.g. CVE auditing)
+- Another benefit is that the user can associate staged channels to the nodes registered (along with an activation key) to associate channels to the nodes of the cluster.
+
+The agent `skuba-update`, running on each node of the cluster, automatically picks up the packages from the associated channels and proceed to automatically update the nodes.
+NOTE: `skuba-update` is **NOT** invoked by Uyuni/SUSE Manager and it is running autonomously on the nodes of the CaaS Platform cluster.
+
+For example, an organization could have a "staging" cluster with a set of staging channels. The operator would cherry-pick Kubernetes updates from the SCC channels into the staging ones and then wait for skuba-update to roll them out into the staging cluster. At a later time, once testing is green inside of the staging cluster, the packages can be promoted to the "production channels" and be rolled out to the production cluster.
+
+It is paramount to state in the documentation that the information presented for the minion is in read-only mode. In the initial iteration, Uyuni/SUSE Manager must point out in the UI that executing actions (package install, upgrade, reboot) might break the cluster if the action are run on a node of the cluster.
+If a user reboots the node, installs a patch or installs a package related to Kubernetes (`kubernetes-kubeadm kubernetes-kubelet kubernetes-client cri-o cni-plugins`) the Kubernetes may break.
+
+NOTE: both of the above actions will not cluster-coordinated, so it is entirely possible that a user assigns different channels, install packages or reboots nodes in a subset of nodes belonging to the same cluster.
+
+## Update the current Kubernetes integration
+
+Under Virtual Host Manager, there is an already Kubernetes integration that supports CaaS Platform version 3.
+The current implementation breaks with CaaS Platform version 4: the Salt runner needs to take into account that CaaS Platform version 4 switched from Docker to CRI-O.
+Uyuni/SUSE Manager must check if the underlying container engine is running Docker (and use the existing Kubernetes integration) or CRI-O (and adapt the current implementation to support CRI-O).
+The Kubernetes implementation that this RFC is targeting is the one in CaaS Platform version 4: 1.16. Other Kubernetes distributions will not be targeted, at least initially.
```

Comment:
It's probably out of the scope of this RFC, but I would be interested to know why you're going so "low level" on the kubernetes cluster to talk straight with the container runtime. I'm pretty sure you can operate at a higher level.

Reply:
What are the tools that we can use at a higher level that we can use?

Reply:
You should just talk with the kubernetes API server using one of the [client libraries](https://kubernetes.io/docs/reference/using-api/client-libraries/).

This will allow you to see all the pods running, look at the containers that are part of it and do any kind of interesting operation. There's no need to talk with the docker daemon running on each node.

Reply:
Now I understand what you are referring to. What I am referring to is that, when Uyuni/SUSE Manager interfaces with the Kubernetes engine, a Salt runner queries the running containers on the cluster: https://bugzilla.suse.com/show_bug.cgi?id=1149741#c0

What I was suggesting to is to check: is the Salt runner returning `docker-pullable` containers? Then we are dealing with a CaaS Platform running Docker. Otherwise, CRI-O is running there.

The sentence is rephrased to clarify the concept. Are we on the same boat on this matter?

Reply:
I looked into the bug. CRI-O (and podman) are more explicit about the source of the image compared to docker.

Docker assumes that, when no registry is specified, the image must come from the Docker Hub (`docker.io`). CRI-O is more explicit about that.

My question to you is: what is SUSE Manager expecting from the output of this salt runner?

Something tells me everything can be solved by changing the code that parses the runner's response.

Reply:
> I looked into the bug. CRI-O (and podman) are more explicit about the source of the image compared to docker.
> 
> Docker assumes that, when no registry is specified, the image must come from the Docker Hub (`docker.io`). CRI-O is more explicit about that.
> 
> My question to you is: what is SUSE Manager expecting from the output of this salt runner?
> 
> Something tells me everything can be solved by changing the code that parses the runner's response.

Yes, the solution would be to modify the Salt runner and add support for images that do not have the `docker-pullable` heading string. I do not think that would be a big deal.